controllers/controllers/workloads/cforg_controller.go

```diff
@@ -78,82 +96,115 @@ func (r *CFOrgReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	if cfOrg.ObjectMeta.DeletionTimestamp != nil && !cfOrg.ObjectMeta.DeletionTimestamp.IsZero() {
-		return ctrl.Result{}, nil
-	}
-
-	var anchor v1alpha2.SubnamespaceAnchor
-	err = r.client.Get(ctx, req.NamespacedName, &anchor)
-	if err != nil {
-		if !k8serrors.IsNotFound(err) {
-			r.log.Error(err, fmt.Sprintf("Error getting SubnamespaceAnchor for CFOrg %s/%s", req.Namespace, req.Name))
+	readyCondition := getConditionOrSetAsUnknown(&cfOrg.Status.Conditions, korifiv1alpha1.ReadyConditionType)
+	if readyCondition == metav1.ConditionUnknown {
+		if err = r.client.Status().Update(ctx, cfOrg); err != nil {
+			r.log.Error(err, fmt.Sprintf("Error when trying to set status conditions on CFOrg %s/%s", req.Namespace, req.Name))
 			return ctrl.Result{}, err
 		}
+	}
 
-		anchorLabels := map[string]string{
-			OrgNameLabel: cfOrg.Spec.DisplayName,
-		}
-		anchor, err = createSubnamespaceAnchor(ctx, r.client, req, cfOrg, anchorLabels)
-		if err != nil {
-			r.log.Error(err, fmt.Sprintf("Error creating SubnamespaceAnchor for CFOrg %s/%s", req.Namespace, req.Name))
-			return ctrl.Result{}, err
-		}
-
-		err = updateStatus(ctx, r.client, cfOrg, metav1.ConditionUnknown)
-		if err != nil {
-			r.log.Error(err, "unable to update CFOrg status")
-			return ctrl.Result{}, err
-		}
+	err = r.addFinalizer(ctx, cfOrg)
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("Error adding finalizer on CFOrg %s/%s", req.Namespace, req.Name))
+		return ctrl.Result{}, err
+	}
 
-		// Requeue to verify subnamespaceanchor is ready
-		return ctrl.Result{RequeueAfter: 100 * time.Millisecond}, nil
+	if isFinalizing(cfOrg) {
+		return finalize(ctx, r.client, r.log, cfOrg, orgFinalizerName)
 	}
 
-	if anchor.Status.State != v1alpha2.Ok {
-		return ctrl.Result{RequeueAfter: 100 * time.Millisecond}, nil
+	labels := map[string]string{korifiv1alpha1.OrgNameLabel: cfOrg.Spec.DisplayName}
+	err = createOrPatchNamespace(ctx, r.client, r.log, cfOrg, labels)
+	if err != nil {
+		r.log.Error(err, fmt.Sprintf("Error creating namespace for CFOrg %s/%s", req.Namespace, req.Name))
+		return ctrl.Result{}, err
 	}
 
 	namespace, ok := getNamespace(ctx, r.client, cfOrg.Name)
 	if !ok {
 		return ctrl.Result{RequeueAfter: 100 * time.Millisecond}, nil
 	}
 
-	err = setCascadingDelete(ctx, r.client, req.Name)
+	err = propagateSecrets(ctx, r.client, r.log, cfOrg, r.packageRegistrySecretName)
```

Comment:
Namespace and related objects propagation is identical in both cforg and cfspace controllers, i.e. we have significant code duplication here. It has been partly addressed by extracting common functions into `shared.go` but from my point of view this is a bit "smelly". According to good OOP principles, using utility entities such as "tools" or "shared" is a smell and indicates code that the developer did not really know where to put.

I would therefore suggest extracting the common namespace handling (create/patch, propagate secrets, roles, rolebindings) in its own "namespace reconciler" that is just injected into the cforg/cfspace controllers. While running the reconciliation loop they could just delegate and if that succeeds, carry on with the org/space specific stuff

Reply:
We split out a chore to follow up on this: https://github.com/cloudfoundry/korifi/issues/1184